### ISSUE TITLE
Add support for the External engine category

### ISSUE THREAD

**Issue by tontsa28** [tontsa28/licheszter#23]
Add support for all endpoints in the External engine category.

**Comment by Malix-Labs** [comment]
Hello!
Do you have a draft implementation for that feature?
Lichess is soon making a new and official external engine tauri-powered software
_And i would personally like to make a new minimal CLI using Rust with your library_

I also have found https://github.com/ion232/lichess-api as an alternative (which partially supports External engine apparently), but I don't know the tradeoffs

**Comment by tontsa28** [comment]
Hi @Malix-Labs!

First, I want to apologize for not replying sooner, I somehow missed your comment and just found it from my notifications tab. I currently don't have any kind of draft implementation made, but I have been thinking about it. Lichess docs say that the API is in alpha and subject to change, which makes me hesitate implementing this feature yet. I assume this might be linked to the tauri software you mentioned? If you have any further details about the future of the external engine feature in general, I would like to hear about them!

TLDR: I don't know if it makes sense to implement the external engine category just yet as there might be a big refactor coming, but it will become relevant at last when the API goes out of alpha. On the other hand, if it will stay in alpha for a long time, it might be already relevant.

And when it ccomes to the `lichess-api` crate, it is indeed an alternative (although I have never tried that crate and don't know the tradeoffs either). What I do know is that this crate is older than `lichess-api`, but shortly after initial release, I didn't develop this crate for like 2 years. That probably motivated the author of `lichess-api` to start their own crate. Now, however, it looks like that crate hasn't been actively maintained for at least 9 months, so I'm not sure what the situation is now. I am planning to continue `licheszter` development at a slow but consistent rate. Depending on how much time I have, my goal is to release v1.0 with all categories supported probably next year (external engine potentially way sooner than that).

**Comment by Malix-Labs** [comment]
No problem for the delay (40 days is not even that much of a delay in open source anyway)!

For the external engine endpoint, it's been alpha (and not really much changing) since nearly 5 years : https://github.com/lichess-org/api/commit/848ecc016581dc1cca137a1e166af2c996778b7f

And thanks for the information about `lichess-api`

**Comment by tontsa28** [comment]
Okay, good to hear that the alpha message has been in the docs for quite long, that means it's probably unlikely they would refactor it in the next year or so (since they haven't done it yet, they will probably not do it now either). That said, I contacted the Lichess devs on Discord regarding this, we'll see if they reply. If they say there is not a major refactor in sight, I might start working on this feature very soon.

**Comment by Malix-Labs** [comment]
I'm also in the discord (\@malix_labs)

You might have further information by reading this thread : https://discord.com/channels/280713822073913354/1463966791603716359

**Comment by Malix-Labs** [comment]
You can also see how https://github.com/fitztrev/lichess-tauri/blob/main/src-tauri/src/lichess.rs implements it (it has a tauri frontend, but the backend is in rust and interacting with the lichess api)

**Comment by tontsa28** [comment]
Yeah, I don't think the implementation should be too difficult. As you might have seen on Discord, trevlar said that there are no plans to do any major updates to the API. That said, I will probably start implementing this feature next. I can't promise anything about the schedule, but I think I should have enough time in the first half of May to get this done (or maybe even sooner, but uni is keeping me busy). I hope that's not too slow for you lol

**Comment by Malix-Labs** [comment]
No of course not, no problem

I was planning to use that part to build a rust cli replacement of  https://github.com/lichess-org/external-engine and then wrap it in a nixpkgs + nixos and home manager service module

I also already opened a PR to package the former in nixpkgs : https://github.com/NixOS/nixpkgs/pull/487931

But it's old, requires the python runtime running 24/7 since lichess doesn't use webhooks or SSE but instead requires client-side polling

**Comment by tontsa28** [comment]
@Malix-Labs If you're still interested, this feature is finally in the main branch and will be released eventually with crate version 0.6. That, however, may take a fairly long time so if you want to test it, I recommend just building the crate from the main branch.

**Comment by Malix-Labs** [comment]
Hello again @tontsa28, I've been extremely busy with work and some open source things, and I've got almost no time to dedicate to chess analysis for now, but I'll tell you when I will have tested it; thanks for the notice!
